packages/cdk8s/lib/metadata.ts

```diff
@@ -0,0 +1,153 @@
+ *
+ * Metadata.of(scope).addLabels({
+ *   app: 'my-app',
+ *   foo: 'bar'
+ * });
+ */
+export class Metadata extends Construct {
+  /**
+   * Returns a `Metadata` object associated with a specific scope.
+   * @param scope The construct scope
+   */
+  public static of(scope: IConstruct): Metadata {
+    return this.tryGetMetadata(scope) ?? new Metadata(scope as Construct, id);
+  }
+
+  /**
+   * Resolves the namespace that should be used for a specific scope.
+   * @param scope The scope to resolve for
+   */
+  public static resolveNamespace(scope?: IConstruct): string | undefined {
+    if (!scope) { return undefined; }
+
+    const resolved = this.lookupMetadata(scope)?.namespace ?? this.resolveNamespace(Node.of(scope).scope);
+
+    // if we resolve to an empty string, then it means the namespace was deleted
+    // so return "undefined".
+    return resolved !== '' ? resolved : undefined;
+  }
+
+  /**
+   * Resolves the labels which should be applied to a scope by merging all
+   * labels from parent scopes.
+   *
+   * @param scope The scope to resolve for
+   */
+  public static resolveLabels(scope?: IConstruct): Record<string, string> {
+    if (!scope) { return { }; }
+
+    return filterUndefined({
+      ...this.resolveLabels(Node.of(scope).scope),
+      ...this.lookupMetadata(scope)?.labels
+    });
+  }
+
+  private static tryGetMetadata(scope: IConstruct): Metadata | undefined {
+    return Node.of(scope).tryFindChild(id) as Metadata | undefined;
+  }
+
+  /**
+   * Returns the first `Metadata` object associated with this scope or a parent scope.
+   *
+   * @param scope The scope to resolve for
+   * @returns the first `Metadata` object defined or undefined if scope is undefined.
+   */
+  private static lookupMetadata(scope?: IConstruct): Metadata | undefined {
+    if (!scope) { return undefined; }
+    return this.tryGetMetadata(scope) ?? this.lookupMetadata(Node.of(scope).scope);
+  }
+
+  private labels?: Labels;
+  private namespace?: string;
+
+  private constructor(scope: Construct, id: string) {
+    super(scope, id);
+  }
+
+  /**
+   * Adds a labels to all the API objects within a scope.
+   *
+   * These labels will be added to all API objects within this scope unless the
+   * same label is set within a narrower scope or at the API object itself.
+   *
+   * If the value is set to `undefined` the label is removed from the scope &
+   * all API objects (equivalent to `removeLabel`).
+   *
+   * @param name The name of the label
+   * @param value The value of the label, or `undefined` to remove the label
+   * from the scope.
+   */
+  public addLabel(name: string, value: string | undefined) {
+    // merge with existing
+    this.labels = sortByKey({
+      ...this.labels,
+      [name]: value
+    });
+  }
+
+  /**
+   * Removes a label from all API objects within the scope.
+   *
+   * @param label The label to remove
+   */
+  public removeLabel(label: string) {
+    this.addLabel(label, undefined);
```

Comment:
Why? An undefined label is akin to no label, isn't it?

Reply:
As far as I understand, the mechanism will have the following heuristic:

- If a label is not configured at all, i.e the key doesn't exist than the label value is taken from the parent.
- If a label is configured but its value is undefined, that value will remain undefined.

Is that correct?

If so, this differentiation feels a little fragile because its easy to write code like this:

```ts
const labels = { something: undefined };

private getLabel(key: string) {
  return labels[key];
};

private applyParentIfNeeded(key: string) {
  const val = getLabel(key);
  if (val === undefined) {
    // lookup parent and apply
    ...
  }
}

applyParentIfNeeded('somethingElse'); // will apply correctly
applyParentIfNeeded('something'); // will apply incorrectly
``` 

My point is that it's easy to write code that returns the same value but has different meaning. I'd prefer that the answer to the question *"was this label explicitly deleted by the user?"* be more explicit.